connectivity/FEATURE_BLE/source/generic/PrivateAddressController.cpp

```diff
@@ -0,0 +1,670 @@
+}
+
+ble_error_t PrivateAddressController::remove_device_from_resolving_list(
+    target_peer_address_type_t peer_address_type,
+    const address_t &peer_identity_address
+)
+{
+    if (is_controller_privacy_supported()) {
+        return queue_remove_device_from_resolving_list(peer_address_type, peer_identity_address);
+    } else {
+        for (auto &entry : _resolving_list) {
+            if (entry.populated &&
+                entry.peer_address_type == peer_address_type &&
+                entry.peer_address == peer_identity_address
+            ) {
+                remove_resolution_entry_from_cache([&](resolution_entry_t& cache_entry) {
+                    return cache_entry.identity == &entry;
+                });
+
+                entry.populated = false;
+
+                restart_resolution_process();
+            }
+        }
+        return BLE_ERROR_NONE;
+    }
+}
+
+ble_error_t PrivateAddressController::clear_resolving_list()
+{
+    if (is_controller_privacy_supported()) {
+        return queue_clear_resolving_list();
+    } else {
+        // Remove entry from the resolving list
+        for (auto &entry : _resolving_list) {
+            entry.populated = false;
+        }
+        // Remove all resolved entries from the cache
+        remove_resolution_entry_from_cache([&](resolution_entry_t& entry) {
+            return entry.identity != nullptr;
+        });
+
+        restart_resolution_process();
+
+        return BLE_ERROR_NONE;
+    }
+}
+
+ble_error_t PrivateAddressController::resolve_address(
+    const address_t &peer_address,
+    bool *resolution_complete,
+    target_peer_address_type_t *retrieved_address_type,
+    const address_t **retrieved_address
+)
+{
+    if (is_controller_privacy_supported()) {
+        return BLE_ERROR_NOT_IMPLEMENTED;
+    }
+
+    // An LRU cache is used, we first traverse the list of resolved address
+    // and return any match
+    auto *entry = _resolved_list;
+    decltype(entry) previous = nullptr;
+    while (entry) {
+        if (entry->address == peer_address) {
+            *resolution_complete = true;
+            // The list contains resolved addresses AND unresolved addresses.
+            // Fill input parameters accordingly
+            if (entry->identity) {
+                *retrieved_address = &entry->identity->peer_address;
+                *retrieved_address_type = entry->identity->peer_address_type;
+            } else {
+                *retrieved_address = nullptr;
+            }
+            // update cache if we're not at the first entry
+            if (previous) {
+                previous->next = entry->next;
+                entry->next = _resolved_list;
+                _resolved_list = entry;
+            }
+            return BLE_ERROR_NONE;
+        }
+        previous = entry;
+        entry = entry->next;
+    }
+
+    // In the case the address has not been resolved, we start the resolution
+    // process.
+    *resolution_complete = false;
```

Comment:
I don't know if it's worth it, now the user has to add extra code, I think it would be simpler to just have one way of returning the result - through the event, the user has to have that codepath anyway as this is not guaranteed to return

Reply:
Good point, that is something I can address.